timemachine/md/exchange/exchange_mover.py

```diff
@@ -0,0 +1,723 @@
+    # compute the number of waters inside the buckyball ligand
+
+    num_ligand_atoms = len(ligand_idxs)
+    num_host_atoms = len(x_t) - num_ligand_atoms
+    host_coords = x_t[:num_host_atoms]
+    ligand_coords = x_t[num_host_atoms:]
+    ligand_centroid = np.mean(ligand_coords, axis=0)
+    count = 0
+    for rj in host_coords:
+        diff = delta_r_np(ligand_centroid, rj, box_t)
+        dij = np.linalg.norm(diff)
+        if dij < threshold:
+            count += 1
+    return count
+
+
+def setup_forcefield(charges):
+    # use a simple charge handler on the ligand to avoid running AM1 on a buckyball
+    ff = Forcefield.load_default()
+    if charges is None:
+        return ff
+
+    q_handle = PrecomputedChargeHandler(charges)
+    q_handle_intra = PrecomputedChargeHandler(charges)
+    q_handle_solv = PrecomputedChargeHandler(charges)
+    return Forcefield(
+        ff.hb_handle,
+        ff.ha_handle,
+        ff.pt_handle,
+        ff.it_handle,
+        q_handle=q_handle,
+        q_handle_solv=q_handle_solv,
+        q_handle_intra=q_handle_intra,
+        lj_handle=ff.lj_handle,
+        lj_handle_solv=ff.lj_handle_solv,
+        lj_handle_intra=ff.lj_handle_intra,
+        protein_ff=ff.protein_ff,
+        water_ff=ff.water_ff,
+    )
+
+
+def image_xvb(initial_state, xvb_t):
+    new_coords = image_frames(initial_state, [xvb_t.coords], [xvb_t.box])[0]
+    return CoordsVelBox(new_coords, xvb_t.velocities, xvb_t.box)
+
+
+def test_exchange():
+    parser = argparse.ArgumentParser(description="Test the exchange protocol in a box of water.")
+    parser.add_argument("--water_pdb", type=str, help="Location of the water PDB", required=True)
+    parser.add_argument(
+        "--ligand_sdf",
+        type=str,
+        help="SDF file containing the ligand of interest. Disable to run bulk water.",
+        required=False,
+    )
+    parser.add_argument("--out_cif", type=str, help="Output cif file", required=True)
+    parser.add_argument(
+        "--ligand_charges",
+        type=str,
+        help='Allowed values: "zero" or "espaloma", required if ligand_sdf is not False.',
+        required=False,
+    )
+
+    parser.add_argument(
+        "--md_steps_per_batch",
+        type=int,
+        help="Number of MD steps per batch",
+        required=True,
+    )
+
+    parser.add_argument(
+        "--mc_steps_per_batch",
+        type=int,
+        help="Number of MC steps per batch",
+        required=True,
+    )
+
+    parser.add_argument(
+        "--insertion_type",
+        type=str,
+        help='Allowed values "targeted" and "untargeted"',
+        required=True,
+    )
+
+    args = parser.parse_args()
+
+    print(" ".join(sys.argv))
+
+    if args.ligand_sdf is not None:
+        suppl = list(Chem.SDMolSupplier(args.ligand_sdf, removeHs=False))
+        mol = suppl[0]
+        if args.ligand_charges == "espaloma":
+            esp = charges.espaloma_charges()  # charged system via espaloma
```

Comment:
Not sure how I'd do this, I can check on the length at best but its just a list of floats (without connectivity information).

Reply:
I've changed this behavior so that the SDF itself is explicit (with documentation in the README):

bb_centered_neutral.sdf C90-buckyball with zero charges (very difficult to insert any water inside)
bb_centered_espaloma.sdf C90-buckyball with espaloma charges (inserts 1-2 waters inside) 

This way, there is zero amiguity about the association of the pre-computed partial charges and *which* SDF file it came from

Reply:
Nice -- I think that's cleaner. (To avoid issues if we ran the same script with a different input `ligand_sdf` path.)